cpp/memilio/utils/multi_index_range.h

```diff
@@ -0,0 +1,162 @@
+/* 
+* Copyright (C) 2020-2023 German Aerospace Center (DLR-SC)
+*
+* Authors: René Schmieding
+*
+* Contact: Martin J. Kuehn <Martin.Kuehn@DLR.de>
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+#ifndef MIO_UTILS_MULTI_INDEX_RANGE_H_
+#define MIO_UTILS_MULTI_INDEX_RANGE_H_
+
+#include "memilio/utils/index.h"
+
+namespace mio
+{
+
+/**
+ * @brief A Range that can be used to iterate over a MultiIndex.
+ * @tparam MultiIndex A type mio::Index<Categories...> for some set of categories.
+ */
+template <class MultiIndex>
+class MultiIndexRange
+{
```

Comment:
I added `make_multi_index_range(index)`, since `make_range(iter_pair&&)` either has priority over `make_range(const index&)` or conflicts with `make_range(index&&)`. It's a long name, but I think it is still fine.

I kept the MultiIndex naming from template arguments elsewhere in the code, but I think IndexRange, IndexIterator, etc. could also work. Which do you think is better? (also @HenrZu)

Reply:
since the multi index is just named `Index`, I think `IndexRange` and `make_index_range` are good.
`IndexIterator` would be wrong, a range is a pair of iterators (in our code as well as in the c++ standard)